File: 프로그래머스/unrated/150369. 택배 배달과 수거하기/택배 배달과 수거하기.py
from heapq import heappop, heappush
def solution(cap, n, deliveries, pickups):
    """
    9:17
    집과 집 사이 거리는 1
    n = 집의 개수
    deliveries 집마다 배달해야될 택배
    pickups 수거해야될 택배들
    """
    answer = 0
    delivery_heap = []
    pickup_heap = []
    for i in range(n):
        if deliveries[i] != 0:
            heappush(delivery_heap, [-i-1, deliveries[i]])
        if pickups[i] != 0:
            heappush(pickup_heap, [-i-1, pickups[i]])
    while delivery_heap or pickup_heap:
        if delivery_heap != [] and pickup_heap != []:
            if delivery_heap[0][0] < pickup_heap[0][0]:
                answer += -2 * delivery_heap[0][0]
            else:
                answer += -2 * pickup_heap[0][0]
            can_carry = cap
            while delivery_heap and can_carry - delivery_heap[0][1] >=0:
                can_carry -= delivery_heap[0][1]
                heappop(delivery_heap)
                
            if can_carry != 0 and delivery_heap != []:
                delivery_heap[0][1] -= can_carry
            can_carry = cap
            while pickup_heap and can_carry - pickup_heap[0][1] >= 0:
                can_carry -= pickup_heap[0][1]
                heappop(pickup_heap)
                
            if can_carry != 0 and pickup_heap != []:
                pickup_heap[0][1] -= can_carry
        elif delivery_heap == []:
            answer += -2 * pickup_heap[0][0]
            can_carry = cap
            while pickup_heap and can_carry - pickup_heap[0][1] >= 0:
                can_carry -= pickup_heap[0][1]
                heappop(pickup_heap)
                
            if can_carry != 0 and pickup_heap != []:
                pickup_heap[0][1] -= can_carry
        else:
            answer += -2 * delivery_heap[0][0]
            can_carry = cap
            while delivery_heap and can_carry - delivery_heap[0][1] >=0:
                can_carry -= delivery_heap[0][1]
                heappop(delivery_heap)
                
            if can_carry != 0 and delivery_heap != []:
                delivery_heap[0][1] -= can_carry
            
        
            
    return answer

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
import java.util.*;

class Solution {
    
    long answer;
    int cap, n;
    
    PriorityQueue<Delivery> deliveryHeap;
    PriorityQueue<Delivery> pickupHeap;
    
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        init(cap, n, deliveries, pickups);
        solve();
        return answer;
    }
    
    void init(int cap, int n, int[] deliveries, int[] pickups) {
        this.cap = cap;
        this.n = n;
        this.answer = 0l;
        deliveryHeap = new PriorityQueue<>(Delivery::compareWithDistance);
        pickupHeap = new PriorityQueue<>(Delivery::compareWithDistance);
        for (int i = 0; i < n; i++) {
            if (deliveries[i] != 0) {
                deliveryHeap.add(new Delivery(i + 1, deliveries[i]));
            }
            if (pickups[i] != 0) {
                pickupHeap.add(new Delivery(i + 1, pickups[i]));
            }
        }
    }
    
    void solve() {
        while(!deliveryHeap.isEmpty() || !pickupHeap.isEmpty()) {
            int delivered = 0;
            int pickuped = 0;
            int maxDistance = 0;
            while (delivered < cap && !deliveryHeap.isEmpty()) {
                Delivery delivery = deliveryHeap.remove();
                delivered += delivery.box;
                maxDistance = Math.max(maxDistance, delivery.distance);
                if (delivered > cap) {
                    deliveryHeap.add(new Delivery(delivery.distance, delivered - cap));
                    delivered = cap;
                }
            }
            
            while (pickuped < cap && !pickupHeap.isEmpty()) {
                Delivery delivery = pickupHeap.remove();
                pickuped += delivery.box;
                maxDistance = Math.max(maxDistance, delivery.distance);
                if (pickuped > cap) {
                    pickupHeap.add(new Delivery(delivery.distance, pickuped - cap));
                    pickuped = cap;
                }
            }
            
            answer += maxDistance * 2;
        }
    }
    
    public static class Delivery {
        int distance;
        int box;
        
        public Delivery(int distance, int box) {
            this.distance = distance;
            this.box = box;
        }
        
        public static int compareWithDistance(Delivery d1, Delivery d2) {
            return Integer.compare(d2.distance, d1.distance);
        }
    }
}